clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1630,6 +1635,52 @@ def GetMemberOp : CIR_Op<"get_member"> {
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// VecElemOp
+//===----------------------------------------------------------------------===//
+
+def VecElemOp : CIR_Op<"vec_elem", [Pure,
+  TypesMatchWith<"type of 'result' matches element type of 'vec'",
+                 "vec", "result",
+                 "$_self.cast<VectorType>().getEltType()">]> {
+
+  let summary = "Extract one element from a vector object";
+  let description = [{
+    The `cir.vec_elem` operation extracts one element from a vector object.
+  }];
+
+  let arguments = (ins CIR_VectorType:$vec, CIR_IntType:$index);
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = [{
+    $vec `[` $index `:` type($index) `]` type($vec) `->` type($result) attr-dict
+  }];
+
+  let hasVerifier = 0;
+}
+
+//===----------------------------------------------------------------------===//
+// VecValue
+//===----------------------------------------------------------------------===//
+
+def VecValueOp : CIR_Op<"vec", [Pure]> {
+
+  let summary = "Create non-const vector value";
```

Comment:
Is the `non-const` part here accurate? If all inputs are cir.const this is semantically a const vector?

Reply:
That wording kind of bugged me when I wrote it.  "non-const" is more accurately "not-necessarily-const", but I don't like that either.  I'll think about it some more and come up with something better, which might just be dropping the "non-const".